src/proof/dot/dot_printer.cpp

```diff
@@ -223,63 +281,117 @@ uint64_t DotPrinter::printInternal(std::ostream& out,
-      break;
-    case PfRule::ASSUME:
-      // a node can belong to more than one view, so these if's must not be
-      // exclusive
-      if (scopeCounter < 2)
-      {
-        classes << " basic";
-        colors << ", color = blue ";
-      }
-      if (inPropositionalView)
-      {
-        classes << " propositional";
-        colors << ", fillcolor = aquamarine4, style = filled ";
-      }
-      break;
-    case PfRule::CHAIN_RESOLUTION:
-    case PfRule::FACTORING:
-    case PfRule::REORDERING:
-      if (inPropositionalView)
-      {
-        classes << " propositional";
-        colors << ", fillcolor = aquamarine4, style = filled ";
-      }
-      break;
-    default: inPropositionalView = false;
-  }
-  classes << " \"";
-  out << classes.str() << colors.str();
   // add number of subchildren
   std::map<const ProofNode*, size_t>::const_iterator it =
       d_subpfCounter.find(pn);
-  out << ", comment = \"{\\\"subProofQty\\\":" << it->second << "}\"";
-  out << " ];\n";
+  out << ", comment = \"{\\\"subProofQty\\\":" << it->second << "}\" ];\n";
+}
 
-  const std::vector<std::shared_ptr<ProofNode>>& children = pn->getChildren();
-  for (const std::shared_ptr<ProofNode>& c : children)
+ProofNodeClusterType DotPrinter::defineProofNodeType(const ProofNode* pn,
+                                                     ProofNodeClusterType last)
+{
+  PfRule rule = pn->getRule();
+  if (isSCOPE(rule))
   {
-    uint64_t childId =
-        printInternal(out, c.get(), pfLet, scopeCounter, inPropositionalView);
-    out << "\t" << childId << " -> " << currentRuleID << ";\n";
+    d_scopesArgs.push_back(pn->getArguments());
   }
 
-  return currentRuleID;
+  // If is the first node
+  if (!d_ruleID)
+  {
+    return ProofNodeClusterType::FIRST_SCOPE;
+  }
+  // If the rule is in the SAT range and the last node was: FF or SAT
+  if (isSat(rule) && last <= ProofNodeClusterType::SAT)
+  {
+    return ProofNodeClusterType::SAT;
+  }
+  // If is a ASSUME
+  if (isASSUME(rule))
+  {
+    if (isInput(pn))
+    {
+      return ProofNodeClusterType::INPUT;
+    }
+    return last;
+  }
+  // the last node was: FS, SAT or CNF
+  if (last <= ProofNodeClusterType::CNF)
+  {
+    // If the rule is in the CNF range
+    if (isCNF(rule))
+    {
+      return ProofNodeClusterType::CNF;
+    }
+    // If the first rule after a CNF is a scope
+    if (isSCOPE(rule))
+    {
+      return ProofNodeClusterType::THEORY_LEMMA;
+    }
+    // Is not a scope
+    return ProofNodeClusterType::PRE_PROCESSING;
+  }
+  // If the last rule was pre processing
+  if (last == ProofNodeClusterType::PRE_PROCESSING)
+  {
+    return ProofNodeClusterType::PRE_PROCESSING;
+  }
+  // If the last rule was theory lemma
+  if (last == ProofNodeClusterType::THEORY_LEMMA)
+  {
+    return ProofNodeClusterType::THEORY_LEMMA;
+  }
+
+  return ProofNodeClusterType::NOT_DEFINED;
+}
+
+inline bool DotPrinter::isInput(const ProofNode* pn)
+{
+  auto& thisAssumeArg = pn->getArguments()[0];
```

Comment:
I'm not sure what happens here with `auto&` and nodes, maybe use `TNode` (the node without reference counting) to make it more clear.